File: python/11-parameter-soup/test/trip_booking_test.py
from src.trip_booking import TripBooking

def test_missing_origin():
    booking = TripBooking()
    result = booking.book_trip(None, "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["error"] == "Missing route"

def test_missing_destination():
    booking = TripBooking()
    result = booking.book_trip("LAX", None, "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["error"] == "Missing route"

def test_missing_departure():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", None, None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["error"] == "Missing departure"

def test_economy_price():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["total"] == 200
    assert result["class"] == "economy"

def test_business_price():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "business", "kosher", "window", None, False, None, False)
    assert result["total"] == 800

def test_first_class_price():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "first", "halal", "window", None, False, None, False)
    assert result["total"] == 2000

def test_save20_promo():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, "SAVE20", False)
    assert result["total"] == 160

def test_save10_promo():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, "SAVE10", False)
    assert result["total"] == 180

def test_insurance():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, True, None, False)
    assert result["total"] == 250

def test_flexible_dates():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, True)
    assert result["total"] == 230

def test_gold_loyalty():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", "GOLD123", False, None, False)
    assert result["total"] == 175

def test_includes_route():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["origin"] == "LAX"
    assert result["destination"] == "NYC"

def test_confirmation_code():
    booking = TripBooking()
    result = booking.book_trip("LAX", "NYC", "2024-01-01", None, "economy", "vegan", "aisle", None, False, None, False)
    assert result["confirmation"].startswith("BK-")

File: javascript/11-parameter-soup/test/trip_booking.test.js
import assert from 'node:assert';
import test from 'node:test';
import { TripBooking } from '../src/trip_booking.js';

test('rejects missing origin', () => {
  const booking = new TripBooking();
  const result = booking.book_trip(null, 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.strictEqual(result.error, 'Missing route');
});

test('rejects missing destination', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', null, '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.strictEqual(result.error, 'Missing route');
});

test('rejects missing departure date', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', null, null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.strictEqual(result.error, 'Missing departure');
});

test('calculates economy price', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.strictEqual(result.total, 200);
  assert.strictEqual(result.class, 'economy');
});

test('calculates business price', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'business', 'kosher', 'window', null, false, null, false);
  assert.strictEqual(result.total, 800);
});

test('calculates first class price', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'first', 'halal', 'window', null, false, null, false);
  assert.strictEqual(result.total, 2000);
});

test('applies SAVE20 promo', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, 'SAVE20', false);
  assert.strictEqual(result.total, 160);
});

test('applies SAVE10 promo', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, 'SAVE10', false);
  assert.strictEqual(result.total, 180);
});

test('adds insurance', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, true, null, false);
  assert.strictEqual(result.total, 250);
});

test('adds flexible dates', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, true);
  assert.strictEqual(result.total, 230);
});

test('applies gold loyalty discount', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', 'GOLD123', false, null, false);
  assert.strictEqual(result.total, 175);
});

test('includes route in result', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.strictEqual(result.origin, 'LAX');
  assert.strictEqual(result.destination, 'NYC');
});

test('includes confirmation code', () => {
  const booking = new TripBooking();
  const result = booking.book_trip('LAX', 'NYC', '2024-01-01', null, 'economy', 'vegan', 'aisle', null, false, null, false);
  assert.ok(result.confirmation.startsWith('BK-'));
});
